Transform terraform hcl var

taherkk opened the pull request.

taherkk committed:
transform hcl variable to json

taherkk committed:
add versions to jq and hcl2json and change their storage paths

taherkk committed:
create bash script to convert hcl variables to json in iacinputdata and create readme for it

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+        exit 1
+    fi
+
+    ARCH=$(uname -m)
```
This will create issue on arch that is `x86_64`, `aarch64`, `armv7l`, and potentially on `i686`. You could create function that detects arch and substitutes these values if necessary.

```bash
_detect_arch() {
  local machine="$(uname -m)"

  declare -A arch_map=(
    ["x86_64"]="amd64"
    ["aarch64"]="arm64"
    ["armv7l"]="arm"
    ["i686"]="i386"
  )

  # Get arch from map if defined, otherwise don't replace
  # For example, in case of arm64, it just echo-es arm64
  echo "${arch_map[$machine]:-$machine}"
}
``` 
and then call it to get proper arch
``` bash
ARCH="$(_detect_arch)"
```

taherkk replied:
fixed

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+
+install_jq(){
+    OS=$(uname -s)
+    if [[ "$OS" == "Darwin" ]]; then
+        OS="macos"
+    elif [[ "$OS" == "Linux" ]]; then
+        OS="linux"
+    else
+        echo "Unsupported OS: $OS"
+        exit 1
+    fi
```
You could also pull this to a function because it's being reused in other places. Something like `_detect_os`, and call it `OS="$(_detect_os)"`

taherkk replied:
fixed

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+    ARCH=$(uname -m)
+
+    JQ_BIN="/tmp/jq"
+    url="https://github.com/jqlang/jq/releases/download/jq-1.8.1/jq-${OS}-${ARCH}"
```
Do we want to always install latest or fixate on `1.8.1`?

taherkk replied:
I find it reliable to fixate. Since if something changes and it breaks you again need to fix it.

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+}
+
+install_hcl2json(){
+    OS=$(uname -s)
+    if [[ "$OS" == "Darwin" ]]; then
+        OS="darwin"
+    elif [[ "$OS" == "Linux" ]]; then
+        OS="linux"
+    else
+        echo "Unsupported OS: $OS"
+        exit 1
+    fi
+
+    ARCH=$(uname -m)
```
Replace with reusable functions mentioned above.

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+
+    HCL2JSON_BIN="/tmp/hcl2json"
+
+    url="https://github.com/tmccombs/hcl2json/releases/download/v0.6.7/hcl2json_${OS}_${ARCH}"
```
Same question, do we lock version or want to use latest?

taherkk replied:
version is always better in my opinion. It won't break things unexpectedly

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+
+
+INPUT_FILE_JSON="$1"
+if [ -z "$INPUT_FILE_JSON" ]; then
```
Let's switch from single `[ ]` to `[[ ]]` since we are already writing everything in `bash`.

Also, if we require file as input, check with:
```
if [[ -f "$INPUT_FILE_JSON" ]]; then
  ...
fi

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+    echo "Usage: $0 <input_file.json>"
+    exit 1
+fi
+
```
I would add global `OS` and `ARCH` variables here before `install_*` functions. This way we only get system info once.

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+# Create a temporary file to store updated objects
+tmpfile=$(mktemp)
+
+> "$tmpfile"
```
The `tmpfile` is created the moment you set it to `$(mkdtemp)`, why are we doing this `> "$tmpfile"`?

taherkk replied:
fixed

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+
+for ((i=0; i<length; i++)); do
+  # Extract ith object
+  obj=$($JQ_BIN ".[$i]" <<<"$json_data")
```
Let's make sure to wrap all `$()` across script into `""`. This will prevent potential word splitting.

hllvc commented on convert_hcl_to_json.sh:
```diff
@@ -0,0 +1,121 @@
+# Combine updated objects into an array and overwrite the original file
+$JQ_BIN -s '.' "$tmpfile" > "$INPUT_FILE_JSON"
+
+rm "$tmpfile" $HCL2JSON_BIN $JQ_BIN
```
We could also include `trap` on top of script which will cover cleanup on SIGINT, EXIT, etc.
 ```bash
trap _cleanup EXIT INT TERM

_cleanup() {
    local files=("$tmpfile" "$HCL2JSON" "$JQ_BIN")
    for file in "${files[@]}"; do
        if [ -f "$file" ]; then
            rm -f "$file"
        fi
    done
}

## And call same function at the end of script for manual trigger
_cleanup
```

hllvc commented on README.md:
```diff
@@ -79,7 +83,7 @@ After completing the export , edit the `sg-payload.json` file to provide tune ea
 
 - Fetch [sg-cli](https://github.com/StackGuardian/sg-cli.git) and set it up locally (documentation present in repo)
-- Run the following commands and pass the `sg-payload.json` as payload (represented below)
+- Run the following commands and pass the `sg-payload.json` as payload (represented belkkkkkkkkkkow)
```
I guess `vim` issue.. Revert to `below`.

hllvc commented:
@taherkk Let's also make sure that all variables and sub-shell executions (`$()`) are quoted to prevent potential word-splitting. Avoid quoting only if you must and expect to split on space or some special characters.

taherkk committed:
fix: typo

hllvc committed:
fix: only convert hcl collection values in convert script

- stop corrupting plain string vars (were rewritten to ${...})
- only parse values that look like hcl objects/lists; pass scalars through
- use jq -r to decode strings instead of manual quote stripping
- map x86_64/aarch64 arch names so downloads work off apple silicon
- write output to temp then mv so input isn't truncated on failure
- curl -fsSL with error checks; set -euo pipefail; trap cleanup
- add brief progress logging
